---
title: "Add CRISPRs Seq Sites"
author: "Haotian Zheng"
date: "3/7/2022"
output: html_document
---
*Need to transform this step into the code line auto.*

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(ggpubr)
library(plyr)
library(reshape2)
library(pheatmap)
library(RColorBrewer)
library(readxl)
library(dplyr)
library(tibble)
library(scales)
library(ggthemes)
library(igraph)
library(psych)
library(vegan)
library(tidyverse)
library(stringr)
do.write <- F
options(scipen = 200)
```

*This document is to add the crisprs sequence position information into the cas protein sequence position information, cause of the CRISPR_Cas.tab file only contains the Cas gene position, so we need to find the crisprs position which match the cas gene, from the crispr_all.tab file.*

### Import the CRISPR_Cas.tab, crisprs_all.tab, and crisprs_orphan.tab, cas_operons_orphan.tab.
```{r}
crispr_cas_origin <- read.table("../Results/CCTyper_Data/all_plasmid_result/CRISPR_Cas.txt", header = T, check.names = F)
crisprs_all_origin <- read.table("../Results/CCTyper_Data/all_plasmid_result/crisprs_all.txt", header = T, check.names = F)
crisprs_orphan <- read.table("../Results/CCTyper_Data/all_plasmid_result/crisprs_orphan.txt", header = T, check.names = F)
cas_operons_orphan <- read.table("../Results/CCTyper_Data/all_plasmid_result/cas_operons_orphan.txt", header = T, check.names = F, sep = "\t")

tempdata <- read.table("../Results/CCTyper_Data/all_plasmid_result/Need_melt_data.txt", header = T, check.names = F, sep = "\t")
```

### Melt the crisprs in the original CRISPR_Cas.tab into one line with all the protein and plasmid positions.
```{r}
# crispr_cas_afsplit <- str_split_fixed(crispr_cas_origin$CRISPRs, ',', 9) %>% as.data.frame()
# crispr_cas_afsplit[crispr_cas_afsplit==""] <- "None" # remove the blank data. We should know that NA have difference with blank data.
# crispr_cas_afsplit2 <- gsub("'", "", crispr_cas_afsplit) %>% as.data.frame()
# crispr_cas_afsplit3 <- gsub("[", "", crispr_cas_afsplit2)
# data.dropna()

tempmelt <- melt(tempdata, id.vars = c("Contig","Operon","Operon_Pos","Prediction"),variable.name = "CRISPRs_group",value.name = c("All_crisprs"))
tempmelt2 <- tempmelt %>% filter(tempmelt$All_crisprs != "None")
write.table(tempmelt2,file = "../Results/CCTyper_Data/all_plasmid_result/tempfile_afmelt.txt",sep = "\t",row.names = FALSE)
```

### Combine the crisprs_all position infomation.
```{r}





```

### Select the min position at the begin, and the max position at the end. Make sure the position have include all the cas and crisprs in it.
```{r}




```



### Fin.