---
title: "System number - Plasmids families"
author: "[Haotian Zheng](https://github.com/marveloushaotian)"
date: "`r Sys.Date()`"
output: html_document
---
**This document analysis:**
*1.Make a bacteria taxonomy data sheet.*  
*2.Draw a tree plot with diff kinds of defense systems in diff bacteria hosts.*  

***

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8, fig.path='Figs/',
                      echo=FALSE, warning=FALSE, message=FALSE) # Need to change the path of the plot.

library(knitr) # evaluate the code, putting the results in a markdown document, which can then be converted to a variety of formats with pandoc.
library(tidyverse) # including ggplot2,dplyr,tidyr,readr,purrr,tibble,stringr,forcats.
# Matrix processing
library(reshape2) # 'melt', 'cast' rebuild database.
library(plyr)
library(data.table)

# Statistic.
library(scales) # generic function whose default method centers and/or scales the columns of a numeric matrix.
library(psych)
library(vegan)
library(randomForest)
library(caret)

# Visualization.
library(igraph)
library(ggrepel) # repel overlapping text labels.
library(ggthemes) # contains extra themes, scales, and geoms, and functions for and related to ggplot2.
library(ggpubr) # facilitates the creation of beautiful ggplot2-based graphs.
library(gg.gap) # Easy to define segments in y-axis for 'ggplot2'.
library(ggbreak) # Effectively Utilize Plotting Space to Deal With Large Datasets and Outliers.
library(RColorBrewer)
library(pheatmap) # A function to draw clustered heatmaps.
library(gggenes) # provides a 'ggplot2' geom and helper functions for draw- ing gene arrow maps.
library(gggenomes)

do.write <- F
options(scipen = 200)
```

#### Import data.
```{r}
combined_data2 <- read.table("../../Results/00.Original_Data/all_combined_data.txt", header = T, check.names = F, sep = "\t")
origin_metadata <- read_tsv("../../Results/00.Original_Data/00.meta_data/plsdb_filtered.tsv")



# metadata_origin <- read_tsv("../../Results/00.Original_Data/00.meta_data/plsdb_filtered.tsv")
# combine_data <- read.table("../../Results/00.Original_Data/all_combined_data.txt",header = T,check.names = F,sep = "\t")
# metadata_origin <- metadata_origin %>% dplyr::rename(Plasmid_ID=ACC_NUCCORE)
# metadata_origin_sel <- metadata_origin[,c(1,11)]
# combine_data_sel <- combine_data[,c(1,2)] %>% unique()
# combine_inc <- left_join(metadata_origin_sel,combine_data_sel)
# combine_inc <- combine_inc %>% dplyr::rename(Group=Defense_System_Type)
# inc_plasmids2 <- combine_inc[grepl("Inc|Col", combine_inc$plasmidfinder), ]
# inc_plasmids2[is.na(inc_plasmids2)] <- "None"
```

***

#### Filter data.
```{r}

```

***

#### Visualization.
```{r}

```

***

##### Fin.